crates/ruff/src/checkers/ast/mod.rs

```diff
@@ -3148,6 +3163,21 @@ where
                     {
                         flake8_pyi::rules::duplicate_union_member(self, expr);
                     }
+
+                    if self.enabled(Rule::UnnecessaryLiteralUnion)
+                        // Avoid duplicate checks if the parent is an `|`
+                        && self.semantic.expr_parent().map_or(true, |parent| {
+                            !matches!(
+                                parent,
+                                Expr::BinOp(ast::ExprBinOp {
+                                    op: Operator::BitOr,
+                                    ..
+                                })
+                            )
+                        })
```

Comment:
I think you can flatten this by using the `matches` directly on the `expr_parent`

```suggestion
                        && matches!(self.semantic.expr_parent(), Some(Expr::BinOp(ast::ExprBinOp { op: Operator::BitOr,                                    .. })
```

... I've probably messed up the parentheses but you get the idea ;)